### ISSUE TITLE
grep causes false positives with SC2016, '\$123' '\$(ab|cd)'

### ISSUE THREAD

**Issue by oittaa** [koalaman/shellcheck#3217]
The documentation states https://www.shellcheck.net/wiki/SC2016:
> ShellCheck tries to increase the signal-to-noise ratio of this warning by ignoring certain well known commands that frequently expect literal dollar signs, such as sh and perl.

It would be nice if it tried to detect at least some common grep patterns.

A simple test case would be checking if there's a literal `$123`.
```sh
#!/bin/sh
head -n 1 /dev/urandom | grep -E '\$123'
``` 
-> SC2016

There are workarounds, but they are quite ugly. Triple backslash within double quotes gets very confusing.
```sh
#!/bin/sh
head -n 1 /dev/urandom | grep -E '\$[1]23'
head -n 1 /dev/urandom | grep -E "\\\$123"
``` 
This passes!

Other common error case goes like this:
```sh
#!/bin/sh
head -n 1 /dev/urandom | grep -E '\$(ab|cd)'
```
-> SC2016

**Comment by kkmuffme** [comment]
It's the same for various other regex tools, e.g. for sed `'s/foo$//'`
